hbase-server/src/main/java/org/apache/hadoop/hbase/snapshot/SnapshotRegionLocator.java

```diff
@@ -0,0 +1,205 @@
+
+  private static final ServerName FAKE_SERVER_NAME =
+    ServerName.parseServerName("www.example.net,1234,1212121212");
+
+  private final TableName tableName;
+  private final TreeMap<byte[], HRegionReplicas> regions;
+
+  private final List<HRegionLocation> rawLocations;
+
+  public static SnapshotRegionLocator create(Configuration conf, TableName table)
+    throws IOException {
+    Path workingDir = new Path(conf.get(getSnapshotManifestDir(table)));
+    FileSystem fs = workingDir.getFileSystem(conf);
+    SnapshotProtos.SnapshotDescription desc =
+      SnapshotDescriptionUtils.readSnapshotInfo(fs, workingDir);
+    SnapshotManifest manifest = SnapshotManifest.open(conf, fs, workingDir, desc);
+
+    TableName tableName = manifest.getTableDescriptor().getTableName();
+    TreeMap<byte[], HRegionReplicas> replicas = new TreeMap<>(Bytes.BYTES_COMPARATOR);
+    List<HRegionLocation> rawLocations = new ArrayList<>();
+
+    for (SnapshotProtos.SnapshotRegionManifest region : manifest.getRegionManifests()) {
+      HBaseProtos.RegionInfo ri = region.getRegionInfo();
+      byte[] key = ri.getStartKey().toByteArray();
+
+      SnapshotHRegionLocation location = toLocation(ri, tableName);
+      rawLocations.add(location);
+      HRegionReplicas hrr = replicas.get(key);
+
+      if (hrr == null) {
+        hrr = new HRegionReplicas(location);
+      } else {
+        hrr.addReplica(location);
+      }
+
+      replicas.put(key, hrr);
+    }
+
+    return new SnapshotRegionLocator(tableName, replicas, rawLocations);
+  }
+
+  private SnapshotRegionLocator(TableName tableName, TreeMap<byte[], HRegionReplicas> regions,
+    List<HRegionLocation> rawLocations) {
+    this.tableName = tableName;
+    this.regions = regions;
+    this.rawLocations = rawLocations;
+  }
+
+  @Override
+  public HRegionLocation getRegionLocation(byte[] row, int replicaId, boolean reload)
+    throws IOException {
+    return regions.floorEntry(row).getValue().getReplica(replicaId);
+  }
+
+  @Override
+  public List<HRegionLocation> getRegionLocations(byte[] row, boolean reload) throws IOException {
+    return List.of(getRegionLocation(row, reload));
+  }
+
+  @Override
+  public void clearRegionLocationCache() {
+
+  }
+
+  @Override
+  public List<HRegionLocation> getAllRegionLocations() throws IOException {
+    return rawLocations;
+  }
+
+  @Override
+  public TableName getName() {
+    return tableName;
+  }
+
+  @Override
+  public void close() throws IOException {
+
+  }
+
+  public static boolean shouldUseSnapshotRegionLocator(Configuration conf, TableName table) {
+    return conf.get(getSnapshotManifestDir(table)) != null;
+  }
+
+  public static void setSnapshotManifestDir(Configuration conf, String dir, TableName table) {
+    conf.set(getSnapshotManifestDir(table), dir);
+  }
+
+  private static String getSnapshotManifestDir(TableName table) {
+    return SNAPSHOT_MANIFEST_DIR_PREFIX
+      + table.getNameWithNamespaceInclAsString().replaceAll("-", "_");
```

Comment:
Just curious. When there is a namespace, do we end up with a directory for the namespace, or do we get a directory with the namespace and tablename concatenated together? I'm guessing it's the latter. That's probably fine, but maybe having the directory-per-namespace would be better. Thoughts?

Reply:
Oh this isn't a directory path, it's a configuration key? My bad.

Reply:
I've noticed that we do have a separate directory for namespaces when running tests. We can verify that in the code [here](https://github.com/apache/hbase/blob/master/hbase-backup/src/main/java/org/apache/hadoop/hbase/backup/HBackupFileSystem.java#L60) as well. The method name is misleading. It should read `getSnapshotManifestDirKey` b/c the method returns the config key for that table to fetch the manifest dir, not the directory itself. 

I can also go ahead and remove call to `replaceAll`. Originally, I had though replacing hyphens for underscores would make sense in the config key, but I don't necessarily think that's true